rfcs/Subworlds.md

```diff
@@ -0,0 +1,139 @@
+
+### Things `Subworlds` would allow 🙂
+- Running systems on subworld `A` on thread `A`, while running the same systems on subworld `B` on thread `B`.
+- Running systems on specific disjoint groups of entities based on some criteria.
+- Running systems on subworld `A` every frame and on subworld `B` every 10 seconds.
+- Helps systems stay clean and not require overuse of _marker-components_ (and possibly _relations_) which can lead to excessive archetypal fracturing which will end up hindering performance. 
+- Give developers the choice to use either a _single_, _global_, world that contains every entity (_note_: for some games, that might makes the most sense and be the most performing), or multiple `Subworlds` if they so desire.
+
+## Reference-level explanation 🧑‍🏫
+
+TODO: expand upon this
+
+```rust
+struct Worlds {
+    subworlds: HashMap<WorldId, World>,
+}
+
+struct World { // aka `Subworld`
+    entities: Entities,
+    components: Components,
+    // more stuff
+}
+```
+
+### Example API (This is not final!) 💻
+```rust
+fn setup(mut commands: Commands) {
+    commands
+        .spawn() // this will spawn the entity into the default world
+        .create_world("World2")
+        .create_world("World3")
+        .set_world("World2") // sets the current world being operated on
+        .spawn(); // will spawn this entity into "World2"
+}
+
+fn on_level_change(mut commands: Commands) {
+    commands
+        .remove_world("World2") // removes the world and all of it's components
+        .add_world("AnotherWorld")
+        .with_system(some_system.system())
+        .unwrap();
+}
+
+fn main() {
+    App::build()
+        // This system will operate on *all* worlds by default
+        .add_system(general_system.system()) 
+        // add this system which only operates on world 2
+        .add_world_system(["World2"], world2_system.system())
+        // add this system which runs on the Default world and World2
+        .add_world_system(["Default", "World2"], two_world_system.system())
+        // provide a RunCriteria
+        .add_world_set(
+            WorldSet::new()
+                .with_run_criteria(some_criteria.system())
+                .with_system(another_system.system())
+                .with_worlds(["World2"])
+        )
+        .run();
+}
+```
+
+- By default, there is a single `World` contained within the `Worlds` (this mirrors the previous behavior of `bevy`).
+- Queries and systems still only operate entities and components from a single `World`.
+  - However, systems can be applied (concurrently) to multiple `Worlds` at a time.
+
+## Drawbacks 🙁
+
+In theory adding a dynamic number of `Subworlds` would add overhead since they must be kept in (something like) a `HashMap<Id, Subworld>`, where as having a single _global_ `World` does not incur this overhead.  
+
+## Rationale and alternatives 📜
+
+- Using _marker-components_ to distinguish entities that reside in disjoint sets. 
```

Comment:
Marker components don't work well for disjoint sets where you don't know the size at compile time. This is the "dynamic component problem".

Relations can get around this to some degree but have different semantics, and "tag components" or the like are not yet implemented or designed at all.